File: vol2/vol2-java-examples/src/main/java/com/heatonresearch/aifh/general/collections/ChooseObject.java
/*
 * Artificial Intelligence for Humans
 * Volume 2: Nature Inspired Algorithms
 * Java Version
 * http://www.aifh.org
 * http://www.jeffheaton.com
 *
 * Code repository:
 * https://github.com/jeffheaton/aifh
 *
 * Copyright 2014 by Jeff Heaton
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * For more information on Heaton Research copyrights, licenses
 * and trademarks visit:
 * http://www.heatonresearch.com/copyright
 */
package com.heatonresearch.aifh.general.collections;

import com.heatonresearch.aifh.randomize.GenerateRandom;
import com.heatonresearch.aifh.randomize.RandomChoice;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

/**
 * This class is used to choose between several objects with a specified probability.
 *
 * @param <T> The type of object to choose from.
 */
public class ChooseObject<T> implements Serializable {
    /**
     * The objects that we are choosing from.
     */
    private final List<ObjectHolder<T>> list = new ArrayList<ObjectHolder<T>>();

    /**
     * The random choose.
     */
    private RandomChoice chooser;

    /**
     * Finalize the structure and set the probabilities.
     */
    public void finalizeStructure() {
        double[] d = new double[size()];
        for (int i = 0; i < size(); i++) {
            d[i] = list.get(i).getProbability();
        }

        this.chooser = new RandomChoice(d);
    }

    /**
     * Add an object.
     *
     * @param probability The probability to choose this object.
     * @param opp         The object to add.
     */
    public void add(double probability, T opp) {
        list.add(new ObjectHolder<T>(opp, probability));
    }

    /**
     * @return The number of objects added.
     */
    public int size() {
        return list.size();
    }

    /**
     * Choose a random object.
     *
     * @param theGenerator  Random number generator.
     * @return The random choice.
     */
    public T pick(GenerateRandom theGenerator) {
        int index = this.chooser.generate(theGenerator);
        return this.list.get(index).getObj();
    }

    /**
     * @return The object to choose from.
     */
    public List<ObjectHolder<T>> getList() {
        return this.list;
    }

    /**
     * CLear all objects from the collection.
     */
    public void clear() {
        this.list.clear();
    }

    /**
     * @return The first object in the list.
     */
    public T pickFirst() {
        return this.list.get(0).getObj();
    }
}

File: vol2/vol2-csharp-examples/AIFH-Vol2-Core/Core/General/Collections/ChooseObject.cs
// Artificial Intelligence for Humans
// Volume 2: Nature-Inspired Algorithms
// C# Version
// http://www.aifh.org
// http://www.jeffheaton.com
//
// Code repository:
// https://github.com/jeffheaton/aifh
//
// Copyright 2014 by Jeff Heaton
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// For more information on Heaton Research copyrights, licenses
// and trademarks visit:
// http://www.heatonresearch.com/copyright
//
using AIFH_Vol2.Core.Randomize;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AIFH_Vol2.Core.General.Collections
{
    /// <summary>
    /// This class is used to choose between several objects with a specified probability.
    /// </summary>
    [Serializable]
    public class ChooseObject<T>
    {
        /// <summary>
        /// The objects that we are choosing from.
        /// </summary>
        private IList<ObjectHolder<T>> _list = new List<ObjectHolder<T>>();

        /// <summary>
        /// The random choose.
        /// </summary>
        private RandomChoice _chooser;

        /// <summary>
        /// Finalize the structure and set the probabilities.
        /// </summary>
        public void FinalizeStructure()
        {
            double[] d = new double[Count];
            for (int i = 0; i < Count; i++)
            {
                d[i] = _list[i].Probability;
            }

            _chooser = new RandomChoice(d);
        }

        /// <summary>
        /// Add an object.
        /// </summary>
        /// <param name="probability">The probability to choose this object.</param>
        /// <param name="opp">The object to add.</param>
        public void Add(double probability, T opp)
        {
            _list.Add(new ObjectHolder<T>(opp, probability));
        }

        /// <summary>
        /// The number of objects added.
        /// </summary>
        public int Count
        {
            get
            {
                return _list.Count;
            }
        }

        /// <summary>
        /// Choose a random object.
        /// </summary>
        /// <param name="theGenerator">Random number generator.</param>
        /// <returns>The random choice.</returns>
        public T Pick(IGenerateRandom theGenerator)
        {
            int index = _chooser.Generate(theGenerator);
            return _list[index].Obj;
        }

        /// <summary>
        /// The object to choose from.
        /// </summary>
        public IList<ObjectHolder<T>> ObjList
        {
            get
            {
                return _list;
            }
        }

        /// <summary>
        /// Clear all objects from the collection.
        /// </summary>
        public void Clear()
        {
            _list.Clear();
        }

        /// <summary>
        /// The first object in the list.
        /// </summary>
        /// <returns></returns>
        public T PickFirst()
        {
            return _list[0].Obj;
        }
    }
}
